src/components/tl/cuda/tl_cuda_topo.c

```diff
@@ -5,172 +5,293 @@
+ucc_tl_cuda_topo_graph_add_link(ucc_tl_cuda_topo_t *topo,
+                                ucc_tl_cuda_topo_node_t *src,
+                                ucc_tl_cuda_topo_node_t *dst)
+{
+    ucc_tl_cuda_topo_link_t *link;
+    ucc_list_for_each(link, &src->link.list_link, list_link) {
+        if (ucc_tl_cuda_topo_device_id_equal(&link->pci_id, &dst->pci_id)) {
+            link->width += 1;
+            return UCC_OK;
+        }
+    }
+    link = (ucc_tl_cuda_topo_link_t*)ucc_malloc(sizeof(*link));
+    if (!link) {
+        tl_error(topo->lib, "failed to allocate topo link");
         return UCC_ERR_NO_MEMORY;
     }
+    link->pci_id = dst->pci_id;
+    link->width  = 1;
+    ucc_list_add_tail(&src->link.list_link, &link->list_link);
+
+    return UCC_OK;
+}
+
+static void ucc_tl_cuda_topo_free_link(ucc_tl_cuda_topo_link_t *link) {
+    return;
+}
+
+static void ucc_tl_cuda_topo_graph_destroy(ucc_tl_cuda_topo_t *topo)
+{
+    int i;
+
+    for (i = 0; i < topo->num_nodes; i++) {
+        ucc_list_destruct(&topo->graph[i].link.list_link,
+                          ucc_tl_cuda_topo_link_t, ucc_tl_cuda_topo_free_link,
+                          list_link);
+    }
+    kh_destroy_inplace(bus_to_node, &topo->bus_to_node_hash);
+    free(topo->graph);
+}
+
+static ucc_status_t ucc_tl_cuda_topo_graph_create(ucc_tl_cuda_topo_t *topo)
+{
+    ucc_status_t status = UCC_OK;
+    cudaError_t cuda_st;
+    char pci_bus_str[MAX_PCI_BUS_ID_STR];
+    nvmlDevice_t nvml_dev;
+    nvmlFieldValue_t nvml_value;
+    nvmlPciInfo_t nvml_pci;
+    nvmlIntNvLinkDeviceType_t nvml_dev_type;
+    ucc_tl_cuda_device_id_t pci_id;
+    ucc_tl_cuda_topo_node_t *node, *peer_node;
+    int num_gpus;
+    int i, num_nvlinks, link;
 
-    p = 0;
-    for (i = 0; i < size; i++) {
-        for (j = 0; j < size; j++) {
-            if (ucc_tl_cuda_topo_is_direct(team, topo, i, j)) {
+    cuda_st = cudaGetDeviceCount(&num_gpus);
+    if ((cuda_st != cudaSuccess) || (num_gpus == 0)){
+        tl_info(topo->lib, "cudaGetDeviceCount failed or no GPU devices found");
+        return UCC_ERR_NO_RESOURCE;
+    }
+    topo->num_nodes = 0;
+    topo->graph = (ucc_tl_cuda_topo_node_t*)ucc_malloc(MAX_PCI_DEVICES *
+                                                       sizeof(*topo->graph),
+                                                       "tl cuda topo graph");
+    if (!topo->graph) {
+        tl_error(topo->lib, "failed to allocate tl cuda topo graph");
+        return UCC_ERR_NO_MEMORY;
+    }
+    kh_init_inplace(bus_to_node, &topo->bus_to_node_hash);
+    pthread_mutex_lock(&nvml_lock);
+    NVMLCHECK_GOTO(nvmlInit_v2(), exit_free_graph, status, topo->lib);
+    nvml_value.fieldId = NVML_FI_DEV_NVLINK_LINK_COUNT;
+    for (i = 0; i < num_gpus; i++) {
+        CUDACHECK_GOTO(cudaDeviceGetPCIBusId(pci_bus_str, MAX_PCI_BUS_ID_STR, i),
+                       exit_nvml_shutdown, status, topo->lib);
+        status = ucc_tl_cuda_topo_get_pci_id(topo->lib, i, &pci_id);
```

Comment:
can we use ucc_tl_cuda_topo_pci_id_from_str() as we already know pci_bus_str ?

Reply:
done